### PR TITLE
feat: Add working-directory attribute

### LINKED ISSUE #2082: Add `[cd(DIR)]` attribute
Allow setting the working directory of a recipe to an arbitrary directory. Good first issue!

### PR DESCRIPTION
**Pull request by bcheidemann**
Closes https://github.com/casey/just/issues/2082

### CODE REVIEW

File: src/recipe.rs [hunk 1]
```diff
@@ -133,11 +133,24 @@ impl<'src, D> Recipe<'src, D> {
...
   fn working_directory<'a>(&'a self, context: &'a ExecutionContext) -> Option<PathBuf> {
-    if self.change_directory() {
-      Some(context.working_directory())
-    } else {
-      None
+    if !self.change_directory() {
+      return None;
     }
+
+    let working_dir = self
```
Comment (casey) [inline src/recipe.rs]: I think this is better implemented as a for loop:

```rust
let working_directory = context.working_directory();

for attribute in self.attributes {
  if let Attribute::WorkingDirectory(dir) = attribute {
    return Some(working_directory.join(dir.cooked);
  }
}

Some(working_directory)
```

Also note the use of `cooked`, which is the contents of the string literal after processing escape sequences.

File: tests/working_directory.rs [hunk 2]
```diff
@@ -331,3 +331,70 @@ file := shell('cat file.txt')
...
+fn attribute() {
+  Test::new()
+    .justfile(
+      r#"
+      [working-directory('bar')]
+      print1:
+        echo "$(basename "$PWD")"
+
+      [working-directory('bar')]
+      [working-directory('baz')]
```
Comment (casey) [inline tests/working_directory.rs]: Duplicate working directory attributes should be an error.

**Review by casey — CHANGES_REQUESTED** [review]
Reviewed, see comments!

File: src/recipe.rs (hunk 1 above)
Comment (bcheidemann) [reply src/recipe.rs]: > I think this is better implemented as a for loop

Yup, the suggested for loop reads better and is more concise. Just curious, is there any other reason you suggested this refactor?

> Also note the use of cooked, which is the contents of the string literal after processing escape sequences.

Good to know - thanks for the info!

**Review by casey — APPROVED** [review]
Sorry for taking so long to get this in! I made `[no-cd]` and `[working-directory]` conflict, so you can't have them both on a recipe. I also added an `expect_file` function to the test runner, which allowed the tests to be simplified, since you can `touch` a file and check that a file in the expected directory was created.

**Comment by bcheidemann** [comment]
Thanks @casey! Looking forward to using this in my projects 😍

**Comment by CoreyLivingston** [comment]
Nice, happy to have this.

**Pull request merged**